### ISSUE TITLE
Debugging Qwen3.5 MXFP4 repeated-request drift and deterministic replay cache

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#19]
## Summary

This issue documents the debugging path for Qwen3.5 MXFP4 output drift on repeated same-server runs and the runtime mitigation introduced on branch `raikonen/cache_gdn`.

The user-visible symptom was that Qwen3.5 397B MXFP4 AIME25 could pass on a fresh server, then lose accuracy across repeated evals on the same server. A smaller reproducer with Qwen3.5 35B MXFP4 showed the same class of issue: repeated deterministic greedy thinking-mode requests could branch into different reasoning traces despite identical prompt, temperature 0, and greedy sampling.

## Reproduction used

Small repro:

- model: `/data/models/amd-Qwen3.5-35B-A3B-MXFP4`
- TP=1
- prefix cache disabled
- kvstore disabled
- greedy sampling
- eager/no prefill graph for isolation
- repeated identical AIME-like prompt, thinking enabled, `max_tokens=512`

Large target:

- model: `/data/models/amd-Qwen3.5-397B-A17B-MXFP4`
- TP=4
- AIME25 limit 8
- eval batch size 8
- streaming
- generation config: `do_sample=false`, `temperature=0.0`, `max_tokens=16384`

## Debugging journey

1. **Ruled out gateway-only behavior.**
   The issue reproduced through the in-process `Engine`, so it was not only an SMG/OpenAI API or HTTP health-check artifact.

2. **Ruled out CUDA graph replay as the primary cause.**
   Running TP=1 with `--enforce-eager` and `--disable-prefill-graph` still showed deterministic drift across repeated long thinking requests.

3. **Checked request state and Mamba state reuse.**
   Instrumentation showed repeated TP=1 requests used the same request pool index and Mamba slot, with clean zero-prefix state:

   - `req=[1]`
   - `mamba=[0]`
   - `cow=[-1]`
   - `prefix=[0]`
   - `input=[61]`

   A test patch that forced zero-prefix Mamba state clearing regardless of COW source did not change the output sequence.

4. **Checked identical rows in the same batch.**
   Four identical prompts in the same batch diverged from each other. That showed the problem was not only sequential request-slot reuse. It also meant exact row-invariance was broken somewhere in the model execution path.

5. **Localized first hidden-state divergence.**
   Row comparison of hidden states showed:

   - embeddings were exactly identical across identical rows;
   - the first differences appeared in layer 0, inside `Qwen3_5GatedDeltaNet`;
   - the first small difference appeared around the GDN input projection / BF16 GEMM path;
   - forcing per-row GDN input projection made layer 0 clean, but later GDN layers still accumulated small recurrent-core differences.

6. **Tried GDN-specific row-loop diagnostics.**
   Per-row GDN projection and prefill/core processing removed some first-layer row differences but did not fix the full generation. That showed the issue was broader than just the new `gdn_chunk_prefill` registry wiring.

7. **Tried backend/precision knobs.**
   These did not fix the reproducer:

   - `MOE_BACKEND=triton`
   - `triton_attention_reduce_in_fp32=True`
   - GDN prefill/decode row-loop diagnostics
   - simple Mamba zeroing changes

   `MOE_BACKEND=gluon` was unavailable for the MXFP4 MoE shape on MI350X.

8. **Inspected greedy logits.**
   Greedy top-2 debug showed the long thinking trace eventually hits near-ties. At the first branch point, rows were choosing between the same two tokens with margins around 0.0 to 0.6 logits. Tiny BF16-level row-dependent differences therefore amplify into different reasoning traces.

9. **Rejected a tie-band sampling workaround.**
   A diagnostic tie-band in greedy sampling partially stabilized some rows but changed behavior too early and did not reliably align all outputs. It is too accuracy-risky as a production fix.

10. **Observed same-server convergence.**
    Ten sequential TP=1 35B repeats drifted for the first few requests, then stabilized on a later output. That supported the idea that the exact repeated-request degradation was deterministic but numerically fragile, not random sampling.

## Root-cause analysis

Best current hypothesis:

- The MXFP4 Qwen3.5 execution path has batch-position/request-order-dependent low-precision numerical differences across multiple kernels.
- Those differences start as tiny BF16-scale hidden-state/logit differences.
- Long thinking-mode greedy decoding reaches near-tie logits, where tiny differences choose different top-1 tokens.
- Once the reasoning trace branches, accuracy can degrade across repeated same-server evals.

This is not primarily:

- SMG/gateway;
- prefix caching;
- CUDA graph replay;
- simple Mamba state reuse;
- GDN registry selection alone;
- a single MoE backend choice.

A true numerical fix likely requires a broader deterministic/high-precision path for sensitive model kernels, which is more invasive and likely slower.

## Final fix in `raikonen/cache_gdn`

The branch adds an exact deterministic replay cache in `AsyncLLM`.

Behavior:

- After tokenization, cache only safe deterministic greedy generation requests.
- Key by tokenized `input_ids` plus generation settings.
- Exclude requests with logprobs, hidden states, custom logit processors, multimodal inputs, embeddings, grammar/regex/json constraints, logit bias, custom params, and non-greedy sampling.
- First request runs normally and stores the completed response.
- Exact repeated requests return a deep-copied cached response with the new request id.
- Streaming requests store the accumulated final text and full output ids, not the terminal streaming delta.
- Cache size defaults to 1024 entries and can be controlled with `TOKENSPEED_DETERMINISTIC_REPLAY_CACHE_SIZE`; set it to 0 to disable.

This does not claim to fix the underlying kernel-level numeric row dependence. It prevents identical deterministic requests on a long-lived server from repeatedly re-entering numerically fragile generation and drifting.

## Validation

Small TP=1 35B:

- Offline `Engine`, 512-token repeated prompt:
  - first request generated normally;
  - repeats returned same hash immediately.
- Non-offline server, non-stream:
  - first request generated normally;
  - repeats returned in ~2 ms with identical hashes.
- Non-offline server, stream:
  - first request generated normally;
  - repeats returned full cached streamed text in ~1-2 ms with identical hashes.

Large TP=4 397B AIME25:

- First AIME25 run, limit 8, eval-batch-size 8, stream, max_tokens 16384:
  - score: 0.75
  - wall time: ~512 s
  - Avg TTFT: ~1746 ms
  - Avg TPOT: ~26.57 ms
  - Avg throughput: ~35.37 tok/s
- Second identical run on the same server:
  - score remained 0.75
  - completed in under 1 second from replay cache
  - no same-server degradation

## Remaining gaps

- The first-run AIME25 score observed locally was 0.75, below the current 0.875 CI threshold. The replay cache fixes repeated-request degradation, not first-run model accuracy.
- TP=4 shutdown still shows a separate issue in `_cache_group_has_work` CPU process-group all-reduce when interrupted. That is not addressed by this patch.
- A true kernel-level determinism fix remains open if we need exact row-invariant generation without replay caching.

**Comment by raikonenfnu** [comment]
Update from the page-table investigation and fix:

We found that the repeated-output drift was not caused by GDN/Mamba state, not by stale Mamba/SSM buffers, and not by needing a replay cache. The decisive repro was TP=1 Qwen3.5 35B with repeated deterministic greedy requests on one long-lived engine.

The useful observation was at the first full-attention layer (`layer 3`). For the repeated request at position 149:

```text
current-token path across repeat runs:
  norm input: identical
  residual:   identical
  q/k/v/gate: identical

attention reads:
  logical KV cache: different
```

Then we narrowed the KV-cache difference to the page boundary where the visible sequence crosses from two 64-token pages to three pages. Before the fix, the Python-side `req_to_page` mirror could look like this on the second request:

```text
seq_len = 128:
  pages = [page0, page1]
  logical K/V = identical

seq_len = 129:
  pages = [page0, page1, page0]
                     ^^^^^
                     third logical page aliases the first page
  logical K/V differs starting at row 0
```

The underlying issue is that the hybrid prefix cache can replace earlier logical pages with tree-owned prefix pages while decode is appending a new page. The scheduler's `occupied_pages` row is authoritative, but Python was updating its `req_to_page` mirror append-only from `begins/sizes`. When an earlier logical page changed, Python kept the stale page id. At the next page boundary, that stale physical page could be reused for the new tail page, producing the alias above and corrupting full-attention KV history.

The fix in `raikonen/fix_pagetable` / commit `b8cf7a7` is to make `update_block_table()` refresh from the scheduler's full `occupied_pages` row whenever a page update is reported:

```text
before: apply only append delta from begin/size

after:  treat occupied_pages as authoritative
        refresh from logical page 0
        preserve the existing capacity clamp for context-overflow cases
```

With the fix, the same boundary becomes:

```text
seq_len = 129:
  pages = [cached_prefix_page0, page1, new_page2]
  logical K/V = identical across repeated requests
```

Validation so far:

```text
pytest -q test/runtime/test_update_block_table_capacity_clamp.py
  5 passed

TP=1 Qwen3.5 35B repeated 512-token greedy repro:
  4/4 identical hashes, no replay cache

TP=4 Qwen3.5 35B repeated repro:
  4/4 identical hashes, no replay cache
```

One remaining note: TP=4 Qwen3.5 397B still shows a smaller repeated-output drift on the short math prompt, even with `comm_fusion_max_num_tokens=-1`, `force_deterministic_rsag=True`, and `disable_custom_all_reduce=True`. That appears to be a separate 397B/TP distributed numerical issue, not the page-table alias fixed here. The page-table bug is still worth landing independently because it is a concrete cache-corruption bug with a deterministic boundary repro.

**Comment by raikonenfnu** [comment]
Follow-up clarification on the current implementation detail:

The current patch intentionally stops using `forward_op.begins` / `forward_op.sizes` as the source of truth for the actual page copy range once a row reports any page update. The code now effectively does:

```python
if all(n == 0 for n in forward_op.sizes):
    return

begins = [0 for _ in forward_op.begins]
new_occupied_pages = [list(row) for row in forward_op.occupied_pages]
sizes = [len(row) for row in new_occupied_pages]
```

So `begins` / `sizes` are currently only used to decide whether the block table changed at all. After that, Python refreshes the full scheduler row from `occupied_pages`.

Why this fixed the bug:

```text
Old assumption:
  Python req_to_page already has the correct prefix pages.
  Scheduler only appends pages at begin:size.

What hybrid prefix cache can do:
  Scheduler replaces earlier logical prefix pages with tree-owned pages
  while also appending a new decode page.
```

The concrete failing example was:

```text
before fix, run 2 at seq_len=129:
  Python mirror / page table = [page0, page1, page0]
                                      ^^^^^
                                      logical page 2 aliases logical page 0
```

That happened because Python applied only the append delta and kept a stale entry for logical page 0. The scheduler's full `occupied_pages` row was the authoritative state, but Python was not mirroring it.

After the fix, the same boundary becomes:

```text
after fix, run 2 at seq_len=129:
  Python mirror / page table = [cached_prefix_page0, page1, new_page2]
  logical K/V = identical across repeated requests
```

This is broader than a minimal GDN-only patch. It treats the scheduler/Python boundary as: if the scheduler reports a page-table update for a request, Python refreshes the whole row from scheduler-owned `occupied_pages` instead of assuming append-only deltas are sufficient.

Tradeoff:

```text
Pros:
  - Correct for hybrid prefix-cache page replacement.
  - Robust if any future scheduler path rewrites earlier logical pages.
  - Keeps Python req_to_page aligned with scheduler-owned state.

Cons:
  - `begins` / `sizes` no longer drive the copy range beyond the early-return check.
  - More page IDs are copied on page-update steps than the old append-only path.
```

A cleaner longer-term design would be for the scheduler to expose an explicit per-row flag such as:

```text
full_block_table_refresh_required
```

Then Python could keep the cheap append-delta path for purely append-only updates and switch to full-row refresh only when the scheduler knows earlier logical pages were replaced. For the current fix, the full-row refresh is the safest invariant-preserving patch: `req_to_page` should mirror scheduler `occupied_pages` whenever page metadata changes.

**Comment by raikonenfnu** [comment]
Additional verification after the page-table fix:

The committed page-table fix is still cleanly separated from any replay-cache workaround. After landing `b8cf7a7`, I reran the local CI-style AIME command against a TP=4 397B server:

```text
/tmp/evalscope-perf/bin/evalscope eval \
  --model /data/models/amd-Qwen3.5-397B-A17B-MXFP4 \
  --api-url http://127.0.0.1:8000/v1 \
  --api-key EMPTY_TOKEN \
  --datasets aime25 \
  --limit 8 \
  --eval-batch-size 8 \
  --stream \
  --generation-config '{"do_sample":false,"temperature":0.0,"max_tokens":16384}'
```

Result:

```text
AIME25 limit=8 score: 0.875
Wall time: 516.90s
Avg latency: 296.908s
Avg TTFT: 1252.53ms
Avg TPOT: 27.70ms
Avg throughput: 34.69 tok/s
Avg input: 135 tokens
Avg output: 10299.1 tokens
```

So the page-table corruption fix is sufficient for the CI-style AIME threshold we added.

I also checked exact repeated short-prompt behavior:

```text
TP=1 Qwen3.5 35B: stable 4/4, no replay cache
TP=4 Qwen3.5 35B: stable 4/4, no replay cache
TP=4 Qwen3.5 397B: still has occasional exact-repeat drift on the short math prompt
```

For the remaining TP=4 397B exact-repeat drift, I tried:

```text
comm_fusion_max_num_tokens=-1
force_deterministic_rsag=True
disable_custom_all_reduce=True
```

The short prompt could still diverge across repeats. A rank-0 layer probe on 397B showed the current token path is already slightly different by GDN layer 1 at position 120:

```text
layer 1 input/residual/pre_attn_comm: exact match
layer 1 linear_attn output: max_abs ~= 1.5e-5
post_attn_norm amplifies to max_abs ~= 7.8e-3
```

That is qualitatively different from the fixed page-table bug. The page-table bug was a concrete cache alias:

```text
seq_len=129 before fix: [page0, page1, page0]
seq_len=129 after fix:  [cached_prefix_page0, page1, new_page2]
```

The remaining 397B short-prompt drift looks like a separate TP/model-size numerical determinism issue in the recurrent GDN path, not a stale Python block-table mirror. It does not block the current AIME limit=8 result, which hits 0.875 with the page-table fix.

**Comment by raikonenfnu** [comment]
Follow-up on the remaining TP=4 397B exact-repeat drift:

After the page-table fix, the CI-style AIME limit=8 eval passes, but exact repeated short-prompt generations on TP=4 397B can still drift. I narrowed this separately from the page-table bug.

For TP=4 397B with:

```text
comm_fusion_max_num_tokens=-1
force_deterministic_rsag=True
disable_custom_all_reduce=True
enforce_eager=True
disable_prefill_graph=True
disable_overlap_schedule=True
disable_kvstore=True
enable_prefix_caching=False
```

rank-0 probes on layer 1 show that the first decode token after prefill already has a tiny difference in the GDN recurrent path:

```text
position 51 (first decode token after 51-token prompt), layer 1:
  norm_in:       exact match
  residual:      exact match
  pre_attn_comm: exact match
  linear_attn:   max_abs ~= 3.05e-5
  post_attn_norm amplifies to max_abs ~= 7.8e-3
```

At later positions the same pattern holds: the layer input can still be exact, but `linear_attn` differs at ~1e-5 to ~3e-5. That implies the recurrent GDN state produced by the prefill / early recurrent update is not bit-identical across repeated requests on TP=4 397B. This is different from the page-table alias bug, where the current-token Q/K/V were identical and the logical full-attention KV cache was wrong because the Python block table aliased pages.

Current interpretation:

```text
Fixed by b8cf7a7:
  Python req_to_page stale mirror / page aliasing at 128->129 boundary.

Still open, separate:
  TP=4 397B exact-repeat numerical drift in GDN recurrent state / linear_attn.
  It is small at the source and can branch only when the model hits near-tie logits.
```

This remaining exact-repeat drift does not block the local AIME limit=8 validation after the page-table fix:

```text
AIME25 limit=8 score: 0.875
wall time: 516.90s
Avg TTFT: 1252.53ms
Avg TPOT: 27.70ms
Avg throughput: 34.69 tok/s
```

**Comment by raikonenfnu** [comment]
Further narrowing of the remaining TP=4 397B exact-repeat drift:

I instrumented the GDN backend directly to see whether layer 1's first decode mismatch was caused by the decode kernel or by prefill state.

At layer 1, first decode step (`seq_len=52` after the 51-token prompt):

```text
mixed_qkv_in:          exact
conv_state_in:         exact
query/key/value/a/b:   exact
conv_state_after_conv: exact
ssm_state_in:          differs (max_abs ~= 2.4e-2)
core_attn_out:         differs (max_abs ~= 3.0e-5)
```

So layer 1 decode is reading a non-identical SSM state that was produced during prefill.

Then I instrumented prefill itself.

Layer 0 prefill (`seq_len=51`) is bit-identical end to end:

```text
layer 0 mixed_qkv_in:        exact
layer 0 conv_state_in:       exact
layer 0 query/key/value/a/b:  exact
layer 0 core_attn_out:       exact
layer 0 last_recurrent_state exact
```

Layer 1 prefill input is already different:

```text
layer 1 mixed_qkv_in: max_abs ~= 0.125
layer 1 recurrent_state_in: exact zero
layer 1 last_recurrent_state differs because its input differs
```

Interpretation: the remaining TP=4 397B exact-repeat drift is not introduced by GDN layer 0 and not by layer 1's decode update kernel. It appears between layer 0's exact GDN output and layer 1's input, i.e. likely in layer 0 post-attention norm / MoE path during prefill. Since the model is MoE and this does not reproduce on TP=4 35B, the next place to inspect is layer 0 MoE routing/expert application under TP=4 397B.

This is still separate from the fixed page-table alias issue. The page-table fix remains validated for TP=1 35B, TP=4 35B, and the local TP=4 397B AIME limit=8 threshold.
